### ISSUE TITLE
[BUG] AuthorizationCodeCredential doesn't work with Confidential Client

### ISSUE THREAD

**Issue by xiaokangz88** [Azure/azure-sdk-for-java#36994]
**Describe the bug**
AuthorizationCodeCredential doesn't work with Confidential Client

***Exception or Stack Trace***
```
com.microsoft.aad.msal4j.MsalInteractionRequiredException: AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.
Trace ID: b8062c5d-0ea7-44e7-8cf3-50657d417100
Correlation ID: 88e31838-d296-453b-9e87-86be4ed4be41
Timestamp: 2023-09-28 07:02:31Z
	at com.microsoft.aad.msal4j.MsalServiceExceptionFactory.fromHttpResponse(MsalServiceExceptionFactory.java:39) ~[msal4j-1.13.9.jar:1.13.9]
	at com.microsoft.aad.msal4j.TokenRequestExecutor.createAuthenticationResultFromOauthHttpResponse(TokenRequestExecutor.java:161) ~[msal4j-1.13.9.jar:1.13.9]
```

**To Reproduce**
1. Create an AuthorizationCodeCredential instance
```java
var tokenCredential = new AuthorizationCodeCredentialBuilder()
                .clientId(CLIENT_ID)
                .clientSecret(CLIENT_SECRET)
                .redirectUrl(REDIRECT_URI)
                .authorizationCode(authCode)
                .build();
```
2. Call the getToken method two times.
```java
TokenRequestContext context = new TokenRequestContext();
context.setScopes(Arrays.asList(SCOPE.split(" ")));
tokenCredential.getToken(context);
tokenCredential.getToken(context);
```

***Code Snippet***
It looks like the code of getToken method only read from the public client's cache, however, the IdenttiyClient saves the token to confidential client's cache if the client secret is provided.
```java
// getToken() method of AuthorizationCodeCrendetial class
if (isCachePopulated) {
    return identityClient.authenticateWithPublicClientCache(request, cachedToken.get())
        .onErrorResume(t -> Mono.empty());
}

// authenticateWithAuthorizationCode() method of IdentityClient class
if (clientSecret != null) {
    acquireToken = getConfidentialClientInstance(request).getValue()
        .flatMap(pc -> Mono.fromFuture(() -> pc.acquireToken(parametersBuilder.build())));
}
```

**Expected behavior**
The access token can be correctly returned from the Confidential Client's cache.

**Setup (please complete the following information):**
 - OS: Windows
 - IDE: IntelliJ
 - Library/Libraries: com.azure:azure-identity:1.10.1
 - Java version: 17
 - App Server/Environment: Tomcat
 - Frameworks: Spring Boot

**Comment by joshfree** [comment]
Thanks for reporting this issue, @xiaokangz88.  @g2vinay could you give an update on this after you've had a chance to review?

**Comment by g2vinay** [comment]
Thanks @xiaokangz88 for reporting this issue.
I will review this soon and provide an update.

**Comment by g2vinay** [comment]
A fix for this is released in v: 1.11.0-beta.2
It will GA in November. 
Try it out, let us know, if anything else is needed.

**Comment by xiaokangz88** [comment]
Thank you @g2vinay, it works now.
